core/metadata.cpp

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "metadata.h"
+#include "exif.h"
+#include "qthelper.h"
+#include <QString>
+#include <QFile>
+#include <QDateTime>
+
+// Fetch unsigned integer of arbitrary length from file or memory in big-endian order.
+// On file error, return default value.
+// This is hackish, but it works.
+// Note: only works for unsigned integer types. This is not tested for!
+template <typename T>
+static inline T getBE(const char *buf_in)
+{
+	constexpr size_t size = sizeof(T);
+	auto buf = reinterpret_cast<unsigned const char *>(buf_in);
+	T ret = 0;
+	for (size_t i = 0; i < size; ++i)
+		ret = (ret << 8) | buf[i];
+	return ret;
+}
+
+template <typename T>
+static inline T getBE(QFile &f, T def=0)
+{
+	constexpr size_t size = sizeof(T);
+	char buf[size];
+	if (f.read(buf, size) != size)
+		return def;
+	return getBE<T>(buf);
+}
+
+static bool parseExif(QFile &f, struct metadata *metadata)
+{
+	f.seek(0);
+	if (getBE<uint16_t>(f) != 0xffd8)
+		return false;
+	for (;;) {
+		switch (getBE<uint16_t>(f)) {
+		case 0xffc0:
+		case 0xffc2:
+		case 0xffc4:
+		case 0xffd0 ... 0xffd7:
+		case 0xffdb:
+		case 0xffdd:
+		case 0xffe0:
+		case 0xffe2 ... 0xffef:
+		case 0xfffe: {
+			quint16 len = getBE<uint16_t>(f);
+			if (len < 2)
+				return false;
+			f.seek(f.pos() + len - 2); // TODO: switch to QFile::skip()
+			break;
+		}
+		case 0xffe1: {
+			quint16 len = getBE<uint16_t>(f);
+			if (len < 2)
+				return false;
+			len -= 2;
+			QByteArray data = f.read(len);
+			if (data.size() != len)
+				return false;
+			easyexif::EXIFInfo exif;
+			if (exif.parseFromEXIFSegment(reinterpret_cast<const unsigned char *>(data.constData()), len) != PARSE_EXIF_SUCCESS)
+				return false;
+			metadata->longitude.udeg = lrint(1000000.0 * exif.GeoLocation.Longitude);
+			metadata->latitude.udeg = lrint(1000000.0 * exif.GeoLocation.Latitude);
+			metadata->timestamp = exif.epoch();
+			return true;
+		}
+		case 0xffda:
+		case 0xffd9:
+			// We expect EXIF data before any scan data
+			return false;
+		default:
+			return false;
+		}
+	}
+}
+
+static bool parseMP4(QFile &f, metadata *metadata)
+{
+	f.seek(0);
+	bool found_ftyp = false;
+	std::vector<uint64_t> atom_stack;
```

Comment:
any reason to use `std::vector` over QVector?
Qt features are preferred over the the standard C++ library.

Reply:
Actually, yes. `QVector` does reference counting, which is neither wanted nor needed here. All I want is a really primitive wrapper around `malloc()`, `realloc()` and `free()` that takes care of `free()`ing  when the variable goes out of scope.

Reply:
i was going to say, "but hey, we don't have `std::vector` anywhere in our codebase, so why introduce it now?". but it looks like we already have it:
```
core/exif.cpp
qt-models/filtermodels.cpp
```

`exit.cpp` is third party code and the other usage was introduced by you and approved by Dirk here:
dd2466f51899aae4

one thing that using `std::vector` does is it breaks our consistency with `QVector` for the developers that are not super familiar with C++, so that's why i would not have accepted it.